### DISCUSSION TITLE
Using MariaDB as Supersets Backend

### DISCUSSION THREAD

**Discussion by CFM90** [apache/superset#18692] [Q&A / Help] [answered]
Hello,

Is it possible to use a MariaDB as Supersets backend? I tested the `PyMySQL` package and used the string:
`SQLALCHEMY_DATABASE_URI = 'mysql+pymysql://root:MyrootPass@127.0.0.1:3306/Superset_Metadata'`

I receive the error:
```
sqlalchemy.exc.ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check 
the manual that corresponds to your MariaDB server version for the right syntax to use near 'offset INTEGER' at line 1")
[SQL: ALTER TABLE datasources ADD COLUMN offset INTEGER]
```
Indicating a problem with the database driver ([see here](https://docs.sqlalchemy.org/en/14/errors.html?highlight=programmingerror#programmingerror))

- I suspected a conflict between Mysql and MariaDB dialect, because the projets continue to diverge. So I wanted to use the MariaDB only option of PyMySQl, but I learned, that Superset requires SQLAlchemy Version to be < 1.4, but the MariaDB dialect in the PyMySQL Package would require SQLAlchemy to be 1.4 and above. So I could not use the MariaDB Only Mode from SQLAlchemy, described [here](https://docs.sqlalchemy.org/en/14/dialects/mysql.html#mariadb-only-mode).
- When using a python3 script I can access the database and retrieve data on my own. So the communication between my virtual env. (in which Superset is installed) and SQLAlchemy is basically working.
- Furthermore when using the default sqlLite DB, I can use the same MariaDB as a datasource using PyMySQL and the mysql sring. 

Do I miss an important part, or is it simply impossible at the moment to use MariaDB as a backend? 

I follow this installation instructions: [Superset from Scratch](https://superset.apache.org/docs/installation/installing-superset-from-scratch/).

**Comment by villebro** [comment, marked as answer]
Unfortunately SQLAlchemy 1.4 is not yet supported, and probably won't be supported in the near term, although upgrading is on the mid-long term roadmap. Having said that, it appears the error is happening due to the column name "offset" being a reserved keyword and not being quoted, which would indicate a bug in the dialect. Can you try running the following on your database and see if they work:

```sql
CREATE TABLE foo(offset INTEGER);
```

and

```sql
CREATE TABLE foo(`offset` INTEGER);
```

If the first one fails and the second one works, then the dialect is not quoting the column name properly.

**Reply by CFM90** [reply]
You are right, the reserved keyword seems to be the problem:
![grafik](https://user-images.githubusercontent.com/24831579/154480586-0fd0e550-d2ab-4c7a-907d-a4b7f60fb757.png)

**Comment by Dwijad** [comment]
Yes, the issue is with the latest version of MariaDB.

There is no keyword quotes issue in MariaDB 10.3

![Screenshot from 2022-05-02 19-53-01](https://user-images.githubusercontent.com/12824049/166251111-6e83b475-6e59-4e64-8a8e-6a299d86855e.png)

**Comment by gervisc** [comment]
seems to work fine for mariadb 11.3. Didn't know i could use pymysql for this.

**Comment by sfirke** [comment]
This is a good question.  I would *not* recommend using MariaDB as a backend for Superset.  A little context here: #26741.  Every couple of months someone opens a GitHub issue or posts in Slack that their Superset is broken or they can't upgrade to a new release because of a problem specific to MariaDB.

Superset is not tested against anything besides MySQL or Postgres, so anyone running another database backend is running the risk of encountering a problem specific to that DB.  And then they will be on their own to fix it.
